### PR TITLE
Tags policy permissions

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

Permissions on tags creation/delete
Recording 2024-07-03 at 12 56 49

Abilities

- `all:host_tags` - All operations on host tags
- `all:cluster_tags` - All operations on cluster tags
- `all:database_tags` - All operations on database tags
- `all:sap_systems_tags` - All operations on sap_systems tags

The Tags component prevent the click on delete/add tag functionality when the abilities are not compatible with the one configured in the component. 
## How was this tested?

Automated tests

### CODE REVIEW

File: test/trento/tags/policy_test.exs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    Enum.each([:cluster, :host, :sap_system, :database], fn tag_resource ->
+      refute Policy.authorize(:add_tag, user, %{tag_resource: tag_resource, resource: Tag})
+    end)
+  end
+
+  test "should not allow delete_tag action on the resource when the user does not have the right ability" do
+    user = %User{abilities: []}
+
+    Enum.each([:cluster, :host, :sap_system, :database], fn tag_resource ->
+      refute Policy.authorize(:delete_tag, user, %{tag_resource: tag_resource, resource: Tag})
```
Comment (arbulu89) [inline test/trento/tags/policy_test.exs:28]: This should be `remove_tag` instead of `delete_tag` and the same in the test description

File: test/trento/tags/policy_test.exs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  alias Trento.Tags.Policy
+  alias Trento.Tags.Tag
+  alias Trento.Users.User
+
+  test "should allow any action when the user has global ability" do
+    user = %User{abilities: [%{resource: "all", name: "all"}]}
+
+    Enum.each([:cluster, :host, :sap_system, :database], fn tag_resource ->
+      assert Policy.authorize(:add_tag, user, %{tag_resource: tag_resource, resource: Tag})
```
Comment (arbulu89) [inline test/trento/tags/policy_test.exs:12]: you could add the `delete_tag` assertion one liner here as well

File: test/trento_web/controllers/v1/host_controller_test.exs [hunk 3]
```diff
@@ -13,31 +13,6 @@ defmodule TrentoWeb.V1.HostControllerTest do
   setup :setup_api_spec_v1
   setup :setup_user
 
-  describe "forbidden routes" do
```
Comment (arbulu89) [inline test/trento_web/controllers/v1/host_controller_test.exs:16]: Let's remove this change. I have already done it in other PR, and it will conflict otherwise.

File: test/trento_web/controllers/v1/tags_controller_test.exs [hunk 4]
```diff
@@ -1,10 +1,71 @@
...
+  import OpenApiSpex.TestAssertions
   import Trento.Factory
+  import Trento.Support.Helpers.AbilitiesTestHelper
   alias Faker.Color
   alias Trento.Tags.Tag
 
+  setup :setup_api_spec_v1
+  setup :setup_user
+
+  describe "forbidden actions" do
```
Comment (arbulu89) [inline test/trento_web/controllers/v1/tags_controller_test.exs:13]: As the others, i would put this at the end of the file as other test files

File: assets/js/lib/test-utils/index.jsx [hunk 5]
```diff
@@ -16,6 +16,9 @@ const middlewares = [];
 const mockStore = configureStore(middlewares);
 
 export const defaultInitialState = {
+  user: {
```
Comment (arbulu89) [inline assets/js/lib/test-utils/index.jsx:19]: This will conflict with my PR as well :see_no_evil:

File: assets/js/pages/ClusterDetails/ClustersList.test.jsx [hunk 6]
```diff
@@ -8,25 +8,58 @@ import { renderWithRouter, withState } from '@lib/test-utils';
...
-      databasesList: {
-        databaseInstances: [],
-      },
-    };
+  describe('tags operations', () => {
+    it('should disable tag creation and deletion if the user abilities are not compatible', async () => {
+      const state = {
+        ...cleanInitialState,
+        clustersList: {
+          clusters: [].concat(
```
Comment (arbulu89) [inline assets/js/pages/ClusterDetails/ClustersList.test.jsx:37]: Just remove the concat. You could directly assign the `buildList` output

File: assets/js/common/Tags/Tags.jsx [hunk 7]
```diff
@@ -15,13 +16,33 @@ const tagValidationDefaultMessage = (
   </>
 );
 
+function TagDeleteButton({ onClick, disabled }) {
```
Comment (arbulu89) [inline assets/js/common/Tags/Tags.jsx:19]: I find strange the `TagDeleteButton`. I would call it `ExistingTag` maybe.

File: assets/js/common/Tags/Tags.jsx [hunk 8]
```diff
@@ -15,13 +16,33 @@ const tagValidationDefaultMessage = (
...
+  return (
+    <span
+      aria-hidden="true"
+      className={classNames('cursor-pointer group-hover:opacity-60', {
```
Comment (arbulu89) [inline assets/js/common/Tags/Tags.jsx:23]: Are the classes correct?
Originally the `ml-2` was on the `span`, but now you have moved it the icon.
Maybe it doesn't change anything, but I got surprised

Comment (CDimonaco) [reply assets/js/common/Tags/Tags.jsx:23]: This is a styling issue because the tooltip need a wrap (otherwise the refs of the library went wrong and the tooltip component crash), so I moved to the icon itself

File: lib/trento/tags/policy.ex [hunk 9]
```diff
@@ -0,0 +1,28 @@
+defmodule Trento.Tags.Policy do
+  @moduledoc """
+  Policy for the Tag resource
+
+  User with the ability all:all can perform any operation on the tags.
+  User with the ability all:<resource_type>_tags can perform any operations on the tags of the permitted resource.
+
+  Resource type can be one of:
+  host
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/tags/policy.ex]: done!

File: test/trento/tags/policy_test.exs (hunk 1 above)
Comment (CDimonaco) [reply test/trento/tags/policy_test.exs]: done thx, for spotting

File: test/trento/tags/policy_test.exs (hunk 2 above)
Comment (CDimonaco) [reply test/trento/tags/policy_test.exs:12]: done

File: assets/js/pages/ClusterDetails/ClustersList.test.jsx (hunk 6 above)
Comment (CDimonaco) [reply assets/js/pages/ClusterDetails/ClustersList.test.jsx]: done

File: assets/js/common/Tags/Tags.jsx (hunk 7 above)
Comment (CDimonaco) [reply assets/js/common/Tags/Tags.jsx]: done

File: test/trento_web/controllers/v1/host_controller_test.exs (hunk 3 above)
Comment (arbulu89) [reply test/trento_web/controllers/v1/host_controller_test.exs]: @CDimonaco I mean, you didn't do any change here as far as I see, just move the `describe` to the bottom, which I did in the latest PR i merged some minutes ago

File: test/trento_web/controllers/v1/host_controller_test.exs [hunk 10]
```diff
@@ -384,4 +384,29 @@ defmodule TrentoWeb.V1.HostControllerTest do
       )
     end
   end
+
+  describe "forbidden routes" do
```
Comment (arbulu89) [inline test/trento_web/controllers/v1/host_controller_test.exs:388]: Just remove this. The exact same test is just above, under `forbidden routes` with some updates

File: assets/js/pages/DatabasesOverview/DatabasesOverview.jsx [hunk 11]
```diff
@@ -20,10 +20,10 @@ function DatabasesOverview({
   databases,
   databaseInstances,
   loading,
-  userAbilities,
   onTagAdd,
   onTagRemove,
   onInstanceCleanUp,
+  userAbilities,
```
Comment (arbulu89) [inline assets/js/pages/DatabasesOverview/DatabasesOverview.jsx:26]: Why this change of order? :see_no_evil: 
Doesn't look better before all the `onX` functions?
Usually, we leave the function props at the end

File: assets/js/pages/HostsList/HostsList.test.jsx [hunk 12]
```diff
@@ -23,6 +23,36 @@ import { filterTable, clearFilter } from '@common/Table/Table.test';
 import HostsList from './HostsList';
 
 describe('HostsLists component', () => {
+  describe('tag operations', () => {
```
Comment (arbulu89) [inline assets/js/pages/HostsList/HostsList.test.jsx:26]: Let's put this describe at the bottom

File: assets/js/pages/SapSystemsOverviewPage/SapSystemsOverview.test.jsx [hunk 13]
```diff
@@ -16,7 +16,37 @@ import { filterTable, clearFilter } from '@common/Table/Table.test';
 
 import SapSystemsOverview from './SapSystemsOverview';
 
+const userAbilities = [{ name: 'all', resource: 'all' }];
+
 describe('SapSystemsOverviews component', () => {
+  describe('tag operations', () => {
```
Comment (arbulu89) [inline assets/js/pages/SapSystemsOverviewPage/SapSystemsOverview.test.jsx:22]: Let's move this to the bottom

**Review by arbulu89 — APPROVED** [review]
Thank you!
PD: We will need to create some ordering standard at some point. You are moving all the props I put at the bottom to the top XD

File: assets/js/pages/SapSystemsOverviewPage/SapSystemsOverview.test.jsx (hunk 13 above)
Comment (CDimonaco) [reply assets/js/pages/SapSystemsOverviewPage/SapSystemsOverview.test.jsx]: done

File: test/trento_web/controllers/v1/host_controller_test.exs (hunk 10 above)
Comment (CDimonaco) [reply test/trento_web/controllers/v1/host_controller_test.exs]: done

File: assets/js/common/Tags/Tags.test.jsx [hunk 14]
```diff
@@ -7,11 +7,34 @@ import '@testing-library/jest-dom';
...
+      <Tags
+        tags={['thetag']}
+        userAbilities={[{ name: 'all', resource: 'resource' }]}
+        tagAdditionPermittedFor={['all:host_tags']}
+        tagDeletionPermittedFor={['all:host_tags']}
+      />
+    );
+
+    expect(screen.queryByText('Add Tag')).toHaveClass('opacity-50');
+    // grab the X
```
Comment (EMaksy) [inline assets/js/common/Tags/Tags.test.jsx:21]: Thanks for the comment !

File: test/trento_web/controllers/v1/tags_controller_test.exs [hunk 15]
```diff
@@ -183,4 +188,60 @@ defmodule TrentoWeb.V1.TagsControllerTest do
       assert 404 == conn.status
     end
   end
+
+  describe "forbidden actions" do
+    test "should not return forbidden on any controller action if the user have the right ability for the tag resource",
```
Comment (EMaksy) [inline test/trento_web/controllers/v1/tags_controller_test.exs:193]: I guess it is a typo ```if the user have``` -->  ```if the user has```

**Review by EMaksy — APPROVED** [review]
Lgtm!
